File: src/java/magma/app/compile/rule/StatefulOrRule.java
package magma.app.compile.rule;

import magma.api.Tuple2;
import magma.api.Tuple2Impl;
import magma.api.collect.Iters;
import magma.api.collect.list.Iterable;
import magma.api.option.Option;
import magma.app.compile.CompileState;
import magma.app.compile.define.Placeholders;

public record StatefulOrRule<T>(Iterable<StatefulRule<T>> rules) implements StatefulRule<T> {
    public static Tuple2<CompileState, String> compileOrPlaceholder(
            CompileState state,
            String input,
            Iterable<StatefulRule<String>> rules
    ) {
        return new StatefulOrRule<String>(rules).apply(state, input).orElseGet(() -> {
            return new Tuple2Impl<CompileState, String>(state, Placeholders.generatePlaceholder(input));
        });
    }

    @Override
    public Option<Tuple2<CompileState, T>> apply(CompileState state, String input) {
        return this.rules.iter()
                .map((StatefulRule<T> statefulRule) -> statefulRule.apply(state, input))
                .flatMap(Iters::fromOption)
                .next();
    }
}

File: src/ts/magma/app/compile/rule/StatefulOrRule.ts
import { StatefulRule } from "../../../../magma/app/compile/rule/StatefulRule";
import { Iterable } from "../../../../magma/api/collect/list/Iterable";
import { CompileState } from "../../../../magma/app/compile/CompileState";
import { Tuple2 } from "../../../../magma/api/Tuple2";
import { Tuple2Impl } from "../../../../magma/api/Tuple2Impl";
import { Placeholders } from "../../../../magma/app/compile/define/Placeholders";
import { Option } from "../../../../magma/api/option/Option";
import { Iters } from "../../../../magma/api/collect/Iters";
export class StatefulOrRule<T> implements StatefulRule<T> {
	rules: Iterable<StatefulRule<T>>;
	constructor (rules: Iterable<StatefulRule<T>>) {
		this.rules = rules;
	}
	static compileOrPlaceholder(state: CompileState, input: string, rules: Iterable<StatefulRule<string>>): Tuple2<CompileState, string> {
		return new StatefulOrRule<string>(rules).apply(state, input).orElseGet(() => {
			return new Tuple2Impl<CompileState, string>(state, Placeholders.generatePlaceholder(input))/*unknown*/;
		})/*unknown*/;
	}
	apply(state: CompileState, input: string): Option<Tuple2<CompileState, T>> {
		return this.rules.iter().map((statefulRule: StatefulRule<T>) => statefulRule.apply(state, input)/*unknown*/).flatMap(Iters.fromOption).next()/*unknown*/;
	}
}
